kore/src/Kore/Step/Step.hs

```diff
@@ -397,19 +397,42 @@ finalizeRulesParallel
     => Pattern (Target variable)
     -> [UnifiedRule (Target variable)]
     -> unifier (Results variable)
-finalizeRulesParallel initial unifiedRules = do
-    let initialCondition = Conditional.withoutTerm initial
-    results <-
-        Foldable.fold <$> traverse (finalizeRule initialCondition) unifiedRules
-    let unifications = MultiOr.make (Conditional.withoutTerm <$> unifiedRules)
-        remainder = Predicate.fromPredicate (Remainder.remainder' unifications)
-    remainders' <- Monad.Unify.gather $ applyRemainder initial remainder
-    return Step.Results
-        { results = Seq.fromList results
-        , remainders =
-            OrPattern.fromPatterns
-            $ Pattern.mapVariables Target.unwrapVariable <$> remainders'
-        }
+finalizeRulesParallel initial unifiedRules =
+    case Rule.checkFunctionLike (term <$> unifiedRules) (term initial) of
+        Left err -> error $ "Rules:\n" <> show (unparseToString . TermLike.mapVariables Target.unwrapVariable . Rule.left . term <$> unifiedRules) <> err
+        _        -> do
+            let initialCondition = Conditional.withoutTerm initial
+            results <-
+                Foldable.fold
+                <$> traverse (finalizeRule initialCondition) unifiedRules
+            let unifications =
+                    MultiOr.make (Conditional.withoutTerm <$> unifiedRules)
+                remainder =
+                    Predicate.fromPredicate (Remainder.remainder' unifications)
+            remainders' <- Monad.Unify.gather $ applyRemainder initial remainder
+            return Step.Results
+                { results = Seq.fromList results
+                , remainders =
+                    OrPattern.fromPatterns
+                    $ Pattern.mapVariables Target.unwrapVariable <$> remainders'
+                }
+
+finalizeRulesSequenceWithCheck
+    ::  forall unifier variable
+    .   ( SimplifierVariable variable
+        , MonadUnify unifier
+        , Log.WithLog Log.LogMessage unifier
+        )
+    => Rule.StepContext
+    -> Pattern (Target variable)
+    -> [UnifiedRule (Target variable)]
+    -> unifier (Results variable)
+finalizeRulesSequenceWithCheck Rule.SimplificationContext initial unifiedRules =
```

Comment:
I don't think we should assume that the simplification context will always apply rules in sequence.